🚀 2단계 - 요금 조회

hhhhhwi opened the pull request:
안녕하세요 성현님!
2단계 PR 올립니다!

이번에는 도메인 레이어에 대한 테스트를 먼저 작성해보면서 설계 -> 구현
의 방식으로 진행하려고 노력은 했으나..

실제로 구현하다보니.. 먼저 작성한 테스트를 거의 갈아엎다시피했습니다 😅 

생각보다 고민해야되는 지점이 많아 수정 커밋이 많아졌네요;

구현하면서 제일 고민되었던 지점은, 경로에 관한 도메인이 Section에 의존적이 되어버렸다는 지점인데요. 
경로에 대한 모든 정보를 Path 객체에 담으려다 보니.. SectionWeightedEdge 객체 인자로 사용하면서 강한 의존성이 생겨버렸습니다. 
지금 생각해보니, Path 라는 경로에 대한 객체를 만들기 보단 정책을 결정하기 위한 조건들을 객체로 만들었어야 했나..? 라는 생각이 좀 드네요;

일단 수정보다는.. 피드백을 받는게 일단 최선일 것 같아 PR 먼저 올립니다..!

이번 PR도 잘 부탁드립니다!

hhhhhwi committed:
docs: readme 작성

hhhhhwi committed:
test : 요금 계산 객체 FareCalculator에 대한 테스트 작성 및 경로 인수 테스트 수정

hhhhhwi committed:
feat: 거리 별 요금 정책에 따른 요금 계산 로직 구현

hhhhhwi committed:
refactor: 경로 별 거리를 구하기 위해 PathFinder 로직 수정
- 지하철 구간의 정보를 담기 위한 SectionWeightedEdge 객체를 생성
- addPath -> SectionWeightedEdge를 추가
- 경로의 총 거리, 시간 등 요금 결정에 필요한 값들을 담은 Path 객체 생성
- getPath 수정

hhhhhwi committed:
refactor: 경로 조회 시, 요금에 대한 응답 값 추가

hhhhhwi committed:
refactor: 경로 조회 시, 조회 기준 상관 없이 거리 및 시간을 함께 응답하도록 수정

boorownie commented on src/test/resources/features/path.feature:
```diff
@@ -27,17 +27,20 @@ Feature: 지하철 경로 관련 기능
     When "학여울역"부터 "개포동역"까지의 최단 "거리" 경로를 조회하면
     Then 경로에 있는 역 목록은 "학여울역,도곡역,개포동역" 순서대로 구성된다
-    And 경로의 최소 거리는 "10"이다
+    And 경로의 소요 거리는 "10"이고 소요 시간은 "9"이다
+    And 이용 요금은 "1250"원 이다
```
기존 인수 테스트 스텝만 추가 👍

boorownie commented on src/main/java/nextstep/subway/domain/Path.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Path {
+    private final List<Long> vertexs;
+    private final List<SectionWeightedEdge> edges;
+    private final double weight;
```
도메인 설계 측면에서 생각해보면 
경로라는 것은 구간의 목록으로 볼 수 있다고 생각합니다.
그런 관점에서 sections나 edges를 가지는게 자연스러워 보입니다.

jgrapht에 대한 의존이 조금 마음에 들지 않으신다고 하면
그래프를 그리는 부분을 추상화 해서 
Path에서 jgrapht를 직접적으로 의존하지 않게 하는 것도 방법일 것 같아요.

boorownie commented on src/main/java/nextstep/subway/domain/DistancePolicy.java:
```diff
@@ -0,0 +1,43 @@
+package nextstep.subway.domain;
+
+public class DistancePolicy implements FarePolicy{
```
빈으로 등록하지 않으므로 단위 테스트에 용이해졌네요 👍

boorownie commented on src/main/java/nextstep/subway/domain/PathFinder.java:
```diff
@@ -1,44 +1,25 @@
 import java.util.List;
-import java.util.Objects;
 
 public abstract class PathFinder {
```
List<Line>를 가지고 있는 어떤 객체(ex. subwayMap)는 경로(Path)객체를 만드는 책임을 가지게 한다면 어떨까요?
PathService에서 PathFinder를 가지고 Path를 만들어주고 있는데
객체 설계적인 측면에서는 조금 자연스럽지 않나요?

boorownie commented on src/test/java/nextstep/subway/unit/PathByDurationFinderTest.java:
```diff
@@ -32,13 +31,13 @@ void setup() {
 	@DisplayName("경로 조회")
-	void getPath() {
-		assertThat(pathFinder.getPath(첫번째_역, 세번째_역)).containsExactly(첫번째_역, 두번째_역, 세번째_역);
+	void getVertex() {
+		assertThat(pathFinder.getPath(첫번째_역, 세번째_역).getVertexs()).containsExactly(첫번째_역, 두번째_역, 세번째_역);
```
vertex라는 용어가 직접 사용되었네요. 혹시 기술적 용어를 드러내었을 때 효과적인 다른 의도가 있으신가요?

boorownie reviewed (changes requested):
고민하시는 부분에 대해 제 생각을 조금 코멘트 남겨보았습니다.
사실 정답이 없는 부분이다보니 제안을 패스하시려면 코멘트 남기고 패스하셔도 되고
수긍되거나 시도해보시려면 반영하셔도 됩니다.

테스트를 갈아 엎다시피했다는 말씀은 두가지로 해석할 수 있는데
하나는 TDD 사이클이 잘 안되었을 수 있지만
이렇게 변화가 큰 경우에는 어짜피 다 갈아엎어야 하는 경우도 있으니 괜찮습니다.

그 외는 잘 구현해주셨으니 코멘트 확인 부탁드려요~

hhhhhwi committed:
refactor: 변수명 수정
- vertexs -> stations : Path는 지하철 경로에 대한 내용이므로 그래프 관련 용어인 vertex보다는 station으로 변경

hhhhhwi committed:
refactor: 지하철의 전체 경로에 대한 WeightedMultigraph를 생성하는 SubwayMap 객체 생성
- 경로 생성 및 최단 경로 찾기에 대한 책임 부여

hhhhhwi committed:
refactor: 그래프를 그리는 SubwayMap 클래스를 통해 Path의 SectionWeightedEdge 의존 제거

hhhhhwi committed:
refactor: 최단 거리 subwayMap으로 조회하도록 PathService 수정

hhhhhwi committed:
refactor: PathFinder 삭제

hhhhhwi commented:
안녕하세요 성현님! 피드백 한 번 반영해보았습니다!

>List를 가지고 있는 어떤 객체(ex. subwayMap)는 경로(Path)객체를 만드는 책임을 가지게 한다면 어떨까요?
PathService에서 PathFinder를 가지고 Path를 만들어주고 있는데
객체 설계적인 측면에서는 조금 자연스럽지 않나요?

> vertex라는 용어가 직접 사용되었네요. 혹시 기술적 용어를 드러내었을 때 효과적인 다른 의도가 있으신가요?

성현님이 남겨주신 위 코멘트들이 피드백을 적용하는 데에 큰 아이디어가 되었는데요..!
제가 Path라는 도메인에 대해서 그래프와 혼동을 하고 있어, vertex나 edge라는 기술용어를 변수명으로 지정하고,
책임이 부자연스럽게 설계가 진행됐던 것 같습니다 😅 

위 리팩토링으로 제가 고민하고 있던 SectionWeightedEdge에 대한 의존도 어느 정도 해결한 것 같아 즐겁게 진행했습니다..!

boorownie commented on src/main/java/nextstep/subway/domain/Path.java:
```diff
@@ -7,9 +7,8 @@
 
 public class Path {
-    private final List<Long> vertexs;
+    private final List<Long> stations;
```
쉬운 용어 활용 👍

boorownie commented on src/main/java/nextstep/subway/domain/SubwayMap.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public abstract class SubwayMap {
```
SubwayMap이라는 새로운 도메인이 나오면서 상태와 행위가 응집될 수 있게 되었네요 👍

boorownie commented on src/main/java/nextstep/subway/domain/SubwayMapByDistance.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.List;
+
+public class SubwayMapByDistance extends SubwayMap {
```
상속을 활용한 중복 코드 제거 👍 
조금 다른 방법을 제안해보자면 상속 대신 조합을 활용해보는 것은 어떨까요?
경로 조회 방법에 대해서 인터페이스로 분리하고 거리 기준, 시간 기준 구현체를 만드는 방식이요!

boorownie commented on src/main/java/nextstep/subway/domain/SubwayMapFactory.java:
```diff
@@ -5,13 +5,13 @@
 
-public class PathFinderFactory {
-	public static PathFinder getPathFinder(List<Section> sections, PathType type) {
+public class SubwayMapFactory {
+	public static SubwayMap getSubwayMap(List<Section> sections, PathType type) {
```
지금도 좋지만! 궁금한 점!
SubwayMap 생성자 대신 Factory를 별도로 만들어 준 이유가 있나요?

boorownie reviewed (approved):
피드백 반영 확인했습니다!
조금 더 객체의 책임과 역할이 구분되는 방향으로 리팩터링을 잘 해주셨어요 👍 
이어서 마지막 단계 진행해주세요~
